### PR TITLE
coord,sql: add new AS OF variants

### LINKED ISSUE #8390: Introduce `AS OF AT LEAST <timestamp>` or equivalent
Our `AS OF <timestamp>` modifier to `SELECT` provides the instruction that a query should be executed at a specific logical time. This can be a bit too specific for folks who want to execute a query at any available time at least `<timestamp>`, without needing to name the specific time. "The first available time at least `<timestamp>`" for example. 

Having to name a specific timestamp can be frustrating, and the best recommendation we have at the moment is to `SELECT .. AS OF ..` and respond to "data no longer available" errors with a `SELECT ..` query.

@mjibson had the take that `AS OF` could be repurposed to just mean this, which I'm personally less excited about, but if no users have bound to this yet that could be available (the term "as of" applies to e.g. as-of joins, appears in SQL specs on temporal queries, and .. either does or doesn't have some baggage attached to it).

### LINKED ISSUE #11188: Spec out the behavior of `AS OF now()`
Forked from "item 2" in #10587.

It is not clear what the behavior of `AS OF now()` should be. Here is an example of a defect in the current behavior, courtesy of @philip-stoev:

```
materialize=> SELECT now(), count(*) FROM s0 AS OF NOW();
            now            | count  
---------------------------+--------
 2022-02-10 12:59:04.72+00 | 100000
...
materialize=> SELECT now(), count(*) FROM s0 AS OF NOW();
ERROR:  Timestamp (1644497944988) is not valid for all inputs: [Antichain { elements: [1644497944999] }]
```

Thoughts from @frankmcsherry:

> This is "in spec" at the moment, but we should be able to fix it. Imo the coordinator should probably not compact things past now() rendering them invalid at the moment. Afaict, the only way to get the "is not valid for all inputs" is to have a query that prescribes a query time, and unfortunately if inputs have been compacted past that time we cannot pick a new time to catch up to (without violating the AS OF constraint). But we can prevent the coordinator from compacting past now().

Additional Slack discussion: https://materializeinc.slack.com/archives/CMHDK0DK8/p1647096901970989

### PR DESCRIPTION
**Pull request by mjibson**
`since` means the greatest `since` of involved sources.

- `AS OF expr` is unchanged.
- `AS OF AT LEAST expr` is like `AS OF expr`, but will bump forward to the `since` if `expr` is before `since`.
- `AS OF EARLIEST IN TIMELINE` is the same as without `AS OF`, but will ignore any linearizability timestamps once they exist.
- `AS OF LATEST IN TIMELINE` is the same as without `AS OF`, but will not advance the linearizability timestamp once it exists.
- without `AS OF` is now prepared to consult and advance linearizability timestamps in the future.
    
`AS OF now()` is a known broken technique that we used to tell people
(when other things were even more broken, so it was sort of useful). It is
now an error.

Fixes #8390
Fixes #11188

### Motivation

  * This PR adds a known-desirable feature.
  * #8390
  * #11188

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - *Breaking change.* Remove support for `AS OF now()`. Internal timestamp generation and selection has improved. This syntax currently often returns incorrect results and has no benefit. Users should remove the `AS OF` clause from their queries.

### CODE REVIEW

**Comment by mjibson** [comment]
Felt inspired to do this. The quality of the inspiration, though, is unclear.

**Comment by mjibson** [comment]
RFAL!

**Comment by mjibson** [comment]
Two test files were removed because they were testing compaction times using AS OF. We should probably test that, but maybe not in testdrive? That stuff feels like motivation for whatever the coordtest replacement is.

**Review by benesch — COMMENTED** [review]
Mostly LGTM, but I'm too sleepy to look at the changes to coord.rs. Maybe @frankmcsherry can look?

File: src/coord/src/coord.rs [hunk 1]
```diff
@@ -3354,22 +3354,31 @@ impl Coordinator {
 
         let since = self.least_valid_read(&id_bundle, compute_instance);
 
+        let mut advance_to_timestamp: Option<Timestamp> = None;
+        let mut advance_to_since = false;
+        let mut advance_to_upper = false;
+        let mut advance_to_table_ts = false;
+        let mut advance_to_linearizability_ts = false;
+        let mut bump_linearizability_ts = false;
```
Comment (frankmcsherry) [inline src/coord/src/coord.rs:3362]: I read this and immediately concluded "this cannot be the best way to write this" not yet knowing what you will write. :D Let me loop back around after I read the implementation.

Related: Having comments atop the block of code would prep expectations about what it does or should do, and help the readers and reviewer!

Comment (frankmcsherry) [reply src/coord/src/coord.rs:3362]: I read it all and do not understand the implementation!

File: src/sql-parser/src/ast/defs/statement.rs [hunk 2]
```diff
@@ -1952,3 +1952,27 @@ impl AstDisplay for NoticeSeverity {
...
 }
 impl_display!(NoticeSeverity);
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum AsOf<T: AstInfo> {
+    At(Expr<T>),
+    AtLeast(Expr<T>),
+    Earliest,
+    Latest,
+}
```
Comment (frankmcsherry) [inline src/sql-parser/src/ast/defs/statement.rs:1962]: We def need doc comments here.

**Review by frankmcsherry — COMMENTED** [review]
I've read the whole thing. Here are some comments!

1. To borrow a phrase of yours, I'm not sure which problem the `EARLIEST` / `LATEST` solves. Neither provides the recency guarantees of linearizability, and maybe that is fine, but if not then what is the guarantee? Is it just that we know internally the shape of the controls, and want to expose them .. for the benefit of users, or of tests? I don't have opposition to them, but I can't really judge them either, and have a lingering suspicion that they might become antipatterns once we get more on the `serializability`, `strong session serializability`, `strong serializability` bandwagon (which I think is coming?).
2. Is `IN TIMELINE` helpful? It looks extraneous, and smells like a modifier, like you might have `IN <SOMETHING ELSE>`. Is that the intent, or is it here for some other reason?
3. Noted at the location, but the timestamp selection code edits were really hard to read, and to a first approximation I didn't. This is I think a function of GH's interleaving edits, but also the code structure not being spelled out ahead of time. If you'd like it iterate on how to structure this, maybe how to pack up the "carried state" in something other than the six bools, I'm game! I don't know that it is wrong, but could maybe use a round of explaining it in person to distill out the concepts?

I think that there is a big pile of recent thinking on "guarantees we want reads to provide" and "mechanisms for providing these guarantees" and I kind of want folks to get in the same room and write these down together. Same comment would go for a "linearizable read" PR that just jams in some additional logic to `determine_timestamp`, to be clear!

I do like the idea of `AS OF AT LEAST <expr>` and `AS OF AT LEAST now()` morally replacing `AS OF now()`, which could still work but which should be expected to often error, which .. might be what you want but at least we have a better thing for you.

File: src/coord/src/coord.rs (hunk 1 above)
Comment (mjibson) [reply src/coord/src/coord.rs:3362]: The "bump" variable here was based on the assumption that we would implement linearizability with a timeline timestamp that queries could advance and other queries would use when determining timestamps. I think that approach is not how we are going to handle things, so this variable probably goes away.

**Comment by mjibson** [comment]
> I'm not sure which problem the EARLIEST / LATEST solves.

One disclaimer here: this PR existed before the linearizability proposal from Gus, so some of it may have outdated ideas.

With our current code they do exactly one thing which is to opt out of using a transaction's timestamp. But we expect in the future that they will do additional things related to linearizability once we have code that can do linearizability. Specifically, based on the proposal by Gus. That proposal will make all SELECT (without AS OF) queries check with upstream sources. EARLIEST will instead not do that check, and just use mz's since/upper to make timestamp choices. LATEST's goal at the time it was discussed in a slack thread was to respect linearizability of other queries, but also not "bump up" other queries to whatever timestamp it got (this was when we thought we would have some timeline timestamp to do linearizability, which is no longer the plan to do linearizability). So I think that LATEST today would be the same as without AS OF, but it escapes the transaction's timestamp. This would mean that LATEST is only useful within transactions when a user wants to do a weird thing? Mostly only useful in tests I think.

> Is IN TIMELINE helpful? It looks extraneous

I concur, but was not the originator of that syntax (it's from a slack thread). Perhaps we have enough quorum to remove it.

> something other than the six bools

Yeah, it's ugly. I considered a, um, struct that had 6 bools instead which felt equally bad.

> "guarantees we want reads to provide" and "mechanisms for providing these guarantees"

The ugly 6 bools are a sort of approach to this. The AS OF variants inform a thing (the unfortunate 6 bools is "the thing" here) about the guarantees they want to provide. Then there are little if blocks for each bool that are the mechanisms for providing each guarantee.

> AS OF AT LEAST now() morally replacing AS OF now()

I no longer think this is good, and am of the opinion that now() should just be disallowed in any AS OF statement, because I don't know what problem it would solve.

**Comment by mjibson** [comment]
@frankmcsherry Do you have a suggestion of how to better organize this logic? One idea that gets rid of the bools is:

```
pub enum QueryWhen {
    /// The peek should occur at the latest possible timestamp that allows the
    /// peek to complete immediately.
    /// peek to complete immediately. Observes and advances linearizability.
    Immediately,
    /// The peek should occur at the timestamp described by the specified
    /// expression.
    ///
    /// The expression may have any type.
    AtTimestamp(MirScalarExpr),
    /// Same as AtTimestamp, and may advance to an available timestamp.
    AtLeastTimestamp(MirScalarExpr),
    // Identical to Immediately, but ignores linearizability of past queries.
    Earliest,
    // Identical to Immediately, but without advancing linearizabiity for future
    // queries.
    Latest,
}

impl QueryWhen {
    fn advance_to_timestamp(&self) -> Option<Timestamp> {
        match self {
            QueryWhen::AtTimestamp(t) | QueryWhen::AtLeastTimestamp(t) => Some(t),
            _ => None,
        }
    }
    fn advance_to_since(&self) -> bool {
        match self {
            QueryWhen::AtTimestamp(_) => false,
            _ => true,
        }
    }
    fn advance_to_upper(&self, uses_tables: bool) -> bool {
        match self {
            QueryWhen::Immediately | QueryWhen::Earliest | QueryWhen::Latest if !uses_tables => {
                true
            }
            _ => false,
        }
    }
    fn advance_to_table_ts(&self, uses_tables: bool) -> bool {
        match self {
            QueryWhen::Immediately | QueryWhen::Earliest | QueryWhen::Latest if uses_tables => true,
            _ => false,
        }
    }
    fn linearizable(&self) -> bool {
        match self {
            QueryWhen::Immediately | QueryWhen::Latest => true,
            _ => false,
        }
    }
}
```

I'd like to get this moving forward, especially because of the upcoming linearizability work that would like syntax to opt out of linearizability (which would be `AS OF EARLIEST` here).

**Comment by frankmcsherry** [comment]
> I'd like to get this moving forward, especially because of the upcoming linearizability work that would like syntax to opt out of linearizability (which would be AS OF EARLIEST here).

I'm sorry I'm not current on the implementation plans for the three consistency properties, and I don't follow the way in which this logic connects them! Like, `Earliest` and `Latest` seem to be unrelated to sequential and strict sequential, or am I super confused?

**Comment by mjibson** [comment]
My understanding is that `EARLIEST` and `LATEST` are unrelated to sequential and strict sequential, they are only related to linearizability. The thought is: `LATEST` does the source timestamp fetching as described in #11262, `EARLIEST` does not. We can choose which of those to do if a user does not specify `AS OF` (probably the same as `LATEST`).

**Comment by mjibson** [comment]
Closing in favor of #12138

**Pull request closed without merging**